### ISSUE TITLE
missing compiler doesn't fall back to the wrap

### ISSUE THREAD

**Issue by stsp** [mesonbuild/meson#12609]
**Describe the bug**
The `[binaries]` section in toolchain config
has this:
```
[binaries]
nasm = 'nasm-segelf'
```

This overrides the nasm compiler requested
in the project() call.
I also do have this nasm-segelf.wrap:
```
[wrap-git]
url = https://github.com/stsp/nasm.git
revision = elf16
depth = 1

[provide]
program_names = nasm-segelf
```
This properly overrides `find_program('nasm-segelf')`.
But I can't get both overrides to work together.
If nasm-segelf is not installed, then nasm
requested via project() call doesn't fall back
to the nasm-segelf wrap. It just says:
`../meson.build:1:0: ERROR: Unknown compiler(s): [['nasm-segelf']]`
and ignores the wrap.
I don't know if it ignores the wrap because
it is called from project(), or because it is
already overridden by `[binaries]` and 2 overrides
do not chain, but in either case the desired
result is not achieved.

**Expected behavior**
nasm should first be translated to nasm-segelf
via the `[binaries]` override, and then it should
fall back to a wrap.

**Comment by dcbaker** [comment]
This is actually a pretty hard problem to solve, because we need to have the compiler available at configure time to run compiler tests and to know argument syntax, etc. It would basically require `meson setup` to stop, run a nested meson + ninja to build the compiler, and then return to configuration using that compiler. There are a handful of other programs we have the same problem with, but off the top of my head I can't remember what they are.

**Comment by dcbaker** [comment]
No, find_program overrides happen at normal build time, we just creat the right dependency chain to ensure that the build works. If you try to use an overridden find_program at configure time you’ll get an error. That tends to to work for find_program, I’m not sure it will work for compilers. Some thought may be required

**Comment by stsp** [comment]
```
Target machine cpu family: x86
Target machine cpu: i386
Program nasm-segelf found: NO
Fallback to subproject nasm-segelf which provides program nasm-segelf

Executing subproject nasm-segelf 

nasm-segelf| Project name: nasm-segelf
nasm-segelf| Project version: 2.17
```
OK I guess I can see what you mean.
The above is what I see at configure time.
Meson noted the override and configured
the wrap project.
But the _use_ of the program is going to
happen during build, whereas the compiler
may be used by itself during configure too.
I see the problem, thanks for an info.

**Comment by stsp** [comment]
However, understanding the problem
doesn't make me very happy about it. :)
I think this problem is serious, and it
would be good to know if the meson
devs at least share the opinion that it
should be solved, one way or another.
For me its a show-stopper currently.

**Comment by stsp** [comment]
Actually there may also be another
problem. Namely: if there is no compiler,
then how would the one ever build the wrap? :)
Seems the solution indeed needs a bit
of brain-storming.

**Comment by bonzini** [comment]
Closing because of what @dcbaker explained. Compilers are pretty special, in that they must be runnable at `meson setup` time (think of `cc.compiles` or `cc.has_argument`).

**Comment by stsp** [comment]
Well, he explained, but there is
still no solution for running the
custom compilers.

**Comment by bonzini** [comment]
You need to preinstall them, there is no alternative. Or use custom targets and do everything by hand if you prefer.
